File: outlierscount.py
import fun
import rpy2.robjects as robjects
from rpy2.robjects.packages import importr
from rpy2.robjects.vectors import FloatVector
import numpy as np
base = importr('base')



def points_on_semicircle(point1, point2, num_points):
        """
        Generates a grid of points for a semicircle on a unit sphere.
        The semicircle lies on the plane defined by point1, point2, and the origin.
        The first point of the grid will be point2, and the last will be its antipodal point.

        Args:
            point1 (list or np.array): A point on the unit sphere, helping to define the plane.
            point2 (list or np.array): The starting point of the semicircle on the unit sphere.
                                        The semicircle will end at its antipodal point (-point2).
            num_points (int): The number of points to generate along the semicircle.

        Returns:
            np.array: A 2D array where each row is a point [x, y, z] on the semicircle.
                      Returns an empty array if the plane cannot be uniquely defined
                      (i.e., point1, point2, and the origin are collinear).
        """
        p1 = np.array(point1, dtype=float)
        p2 = np.array(point2, dtype=float)

        # 1. Normalize the input points (ensure they are on the unit sphere)
        p1 = p1 / np.linalg.norm(p1)
        p2 = p2 / np.linalg.norm(p2)

        # 2. Calculate the normal vector of the plane containing the semicircle
        # This plane is defined by point1, point2, and the origin.
        normal_vector = np.cross(p1, p2)

        # Check if points defining the plane are collinear with the origin
        if np.linalg.norm(normal_vector) < 1e-9:
            print("Warning: Input points (point1, point2) are collinear with the origin. Cannot form a unique plane.")
            return np.array([])
        normal_vector = normal_vector / np.linalg.norm(normal_vector)

        # 3. The starting point for rotation is p2.
        # We will rotate p2 around the normal_vector from 0 to pi radians.
        angles = np.linspace(0, np.pi, num_points)
        test1= p2 * np.cos(angles[1]) + \
                            np.cross(normal_vector, p2) * np.sin(angles[1]) + \
                            normal_vector * np.dot(normal_vector, p2) * (1 - np.cos(angles[1]))
        test2= p2 * np.cos(angles[-2]) + \
                            np.cross(normal_vector, p2) * np.sin(angles[-2]) + \
                            normal_vector * np.dot(normal_vector, p2) * (1 - np.cos(angles[-2]))

        semicircle_points = []
        if np.sign(np.dot(test2,point1))==np.sign(np.dot(test2,point2)) and np.sign(np.dot(test1,point1))==np.sign(np.dot(test1,point2)) and np.sign(np.dot(test1,point1))==np.sign(np.dot(test2,point1)) and np.sign(np.dot(test1,point2))==np.sign(np.dot(test2,point2)):
            for angle in angles:
                # Rodrigues' rotation formula: v_rot = v * cos(theta) + (k x v) * sin(theta) + k * (k . v) * (1 - cos(theta))
                # Here, v is p2, k is normal_vector, and theta is angle
                rotated_point = p2 * np.cos(angle) + \
                                np.cross(normal_vector, p2) * np.sin(angle) + \
                                normal_vector * np.dot(normal_vector, p2) * (1 - np.cos(angle))
                semicircle_points.append(rotated_point)
        else:
            for angle in angles:
                rotated_point = p2 * np.cos(-angle) + \
                                np.cross(normal_vector, p2) * np.sin(-angle) + \
                                normal_vector * np.dot(normal_vector, p2) * (1 - np.cos(-angle))
                semicircle_points.append(rotated_point)

        return np.array(semicircle_points)




def outliers_count(data, weights=None, dist="arc", a=0.99, borderdist="mean", res=500):
    data = np.asarray(data)
    # x, y and z values from the data
    data_x = data[:, 0]
    data_y = data[:, 1]
    data_z = data[:, 2]
    # loading the data into R instance
    fv_x = FloatVector(data_x)
    fv_y = FloatVector(data_y)
    fv_z = FloatVector(data_z)
    robjects.globalenv['x'] = fv_x
    robjects.globalenv['y'] = fv_y
    robjects.globalenv['z'] = fv_z

    if weights=None:
        weights = np.ones(len(data))
    weights = np.asarray(weights)
    fv_weights = FloatVector(weights)
    robjects.r("data=cbind(x,y,z)")
    robjects.globalenv['weights'] = fv_weights
    # calculate the depth of the datapoints
    datadepth = robjects.r("ahD(data,weights)")
    # finding the depth of the border of the bag
    borderdepth = np.median(datadepth)
    nonbagpoints= [data[i] for i in range(len(data_x)) if datadepth[i] < borderdepth]
    datadict = dict()
    for key in datadepth:
        datadict[key] = []
        indices = [i for i, x in enumerate(datadepth) if x == key]
        for i in indices:
            datadict[key].append(data[i])
    if len(datadict[max(datadepth)]) == 1:
        mid = datadict[max(datadepth)][0]
    else:
        midmean = np.mean(datadict[max(datadepth)], axis=0)
        mid = midmean / np.linalg.norm(midmean)

    phi, theta = np.mgrid[0:np.pi:res * 1j, 0:2 * np.pi:res * 1j]
    x_sphere = np.sin(phi) * np.cos(theta)
    y_sphere = np.sin(phi) * np.sin(theta)
    z_sphere = np.cos(phi)

    rotation_matrix = fun.ortho_matrix(mid)

    x_sphere_rotated = (rotation_matrix[0, 0] * x_sphere +
                        rotation_matrix[0, 1] * y_sphere +
                        rotation_matrix[0, 2] * z_sphere)

    y_sphere_rotated = (rotation_matrix[1, 0] * x_sphere +
                        rotation_matrix[1, 1] * y_sphere +
                        rotation_matrix[1, 2] * z_sphere)

    z_sphere_rotated = (rotation_matrix[2, 0] * x_sphere +
                        rotation_matrix[2, 1] * y_sphere +
                        rotation_matrix[2, 2] * z_sphere)

    surface_colors_data = x_sphere_rotated.copy()
    for i in range(len(x_sphere)):
        x_spherefv = FloatVector(x_sphere_rotated[i])
        y_spherefv = FloatVector(y_sphere_rotated[i])
        z_spherefv = FloatVector(z_sphere_rotated[i])
        robjects.globalenv['x'] = x_spherefv
        robjects.globalenv['y'] = y_spherefv
        robjects.globalenv['z'] = z_spherefv
        robjects.r("coords=cbind(x,y,z)")
        surface_colors_data[i] = np.where(np.array(robjects.r("ahD(data,weights,coords)")) >= borderdepth,
                                          1, 0)
    # finding the borders of the bag
    borders = []
    for i in range(len(x_sphere_rotated)):
        borders.append([np.where(surface_colors_data[:, i] == 0)[0][0] - 1, i])
    middist = x_sphere_rotated.copy()
    for i in range(len(x_sphere_rotated)):
        for j in range(len(x_sphere_rotated[i])):
            if dist == "arc":
                middist[i, j] = fun.arcdist([x_sphere_rotated[i, j], y_sphere_rotated[i, j], z_sphere_rotated[i, j]],
                                            mid)
            elif dist == "cos":
                middist[i, j] = fun.cosdist([x_sphere_rotated[i, j], y_sphere_rotated[i, j], z_sphere_rotated[i, j]],
                                            mid)
            else:
                middist[i, j] = fun.chorddist([x_sphere_rotated[i, j], y_sphere_rotated[i, j], z_sphere_rotated[i, j]],
                                              mid)
    # finding the maximal distance from the center of the bagplot to the border of the bag
    if borderdist == "max":
        borderd = max([middist[b[0], b[1]] for b in borders])
    if borderdist == "mean":
        borderd = np.mean([middist[b[0], b[1]] for b in borders])
    if dist == "arc":
        factor = fun.arcMF(fun.kappaarc(borderd),a)
    elif dist == "cos":
        factor = fun.cosMF(fun.kappacos(borderd),a)
    elif dist == "chord":
        factor = fun.chordMF(fun.kappachord(borderd),a)
    else:
        print("Wrong format of dist")
        return
    num_outliers=0
    for point in nonbagpoints:
        grid = points_on_semicircle(point, mid, res)
        grid_xfv = FloatVector(grid[:,0])
        grid_yfv = FloatVector(grid[:,1])
        grid_zfv = FloatVector(grid[:,2])
        robjects.globalenv['x'] = grid_xfv
        robjects.globalenv['y'] = grid_yfv
        robjects.globalenv['z'] = grid_zfv
        robjects.r("coords=cbind(x,y,z)")
        bagornot =np.where(np.array(robjects.r("ahD(data,weights,coords)")) >= borderdepth,1, 0)
        borderindex = np.where(bagornot == 0)[0][0] - 1

        if dist == "arc":
            borderd = fun.arcdist(grid[borderindex], mid)
            loopd = factor[0] * borderd
            pointdist = fun.arcdist(point, mid)
        elif dist == "cos":
            borderd = fun.cosdist(grid[borderindex], mid)[0]
            loopd = factor[0] * borderd
            pointdist = fun.cosdist(point, mid)
            borderd = fun.chorddist(grid[borderindex], mid)
        elif dist == "chord":
            loopd = factor[0] * borderd
            pointdist = fun.chorddist(point, mid)
        else:
            print("Wrong format of dist")
            return

        if pointdist>loopd:
            num_outliers+=1



    return num_outliers

File: R version/outlierscount.r
source("fun.r")

# Function to generate points on a semicircle
points_on_semicircle <- function(point1, point2, num_points) {
  p1 <- as.numeric(point1)
  p2 <- as.numeric(point2)

  # Normalize the input points
  p1 <- p1 / sqrt(sum(p1^2))
  p2 <- p2 / sqrt(sum(p2^2))

  # Calculate the normal vector of the plane
  normal_vector <- c(
    p1[2] * p2[3] - p1[3] * p2[2],
    p1[3] * p2[1] - p1[1] * p2[3],
    p1[1] * p2[2] - p1[2] * p2[1]
  )

  # Check if collinear
  if (sqrt(sum(normal_vector^2)) < 1e-9) {
    warning("Input points are collinear with the origin. Cannot form a unique plane.")
    return(matrix(numeric(0), nrow = 0, ncol = 3))
  }
  normal_vector <- normal_vector / sqrt(sum(normal_vector^2))

  # Angles from 0 to pi
  angles <- seq(0, pi, length.out = num_points)

  # Test points
  cross_prod <- function(a, b) {
    c(a[2]*b[3] - a[3]*b[2], a[3]*b[1] - a[1]*b[3], a[1]*b[2] - a[2]*b[1])
  }

  test1 <- p2 * cos(angles[2]) +
           cross_prod(normal_vector, p2) * sin(angles[2]) +
           normal_vector * sum(normal_vector * p2) * (1 - cos(angles[2]))
  test2 <- p2 * cos(angles[num_points - 1]) +
           cross_prod(normal_vector, p2) * sin(angles[num_points - 1]) +
           normal_vector * sum(normal_vector * p2) * (1 - cos(angles[num_points - 1]))

  semicircle_points <- list()

  # Check signs
  sign_test1_p1 <- sign(sum(test1 * p1))
  sign_test1_p2 <- sign(sum(test1 * p2))
  sign_test2_p1 <- sign(sum(test2 * p1))
  sign_test2_p2 <- sign(sum(test2 * p2))

  if (sign_test2_p1 == sign_test2_p2 && sign_test1_p1 == sign_test1_p2 &&
      sign_test1_p1 == sign_test2_p1 && sign_test1_p2 == sign_test2_p2) {
    for (angle in angles) {
      rotated_point <- p2 * cos(angle) +
                       cross_prod(normal_vector, p2) * sin(angle) +
                       normal_vector * sum(normal_vector * p2) * (1 - cos(angle))
      semicircle_points <- c(semicircle_points, list(rotated_point))
    }
  } else {
    for (angle in angles) {
      rotated_point <- p2 * cos(-angle) +
                       cross_prod(normal_vector, p2) * sin(-angle) +
                       normal_vector * sum(normal_vector * p2) * (1 - cos(-angle))
      semicircle_points <- c(semicircle_points, list(rotated_point))
    }
  }

  return(do.call(rbind, semicircle_points))
}

# Main outliers_count function
outliers_count <- function(data, weights = NULL, dist = "arc", a = 0.99, borderdist = "mean", res = 500) {
  data <- as.matrix(data)
  data_x <- data[, 1]
  data_y <- data[, 2]
  data_z <- data[, 3]

  if (is.null(weights)) {
    weights <- rep(1, nrow(data))
  }
  weights <- as.numeric(weights)

  # Calculate depths
  datadepth <- ahD(data, weights)

  # Border depth
  borderdepth <- median(datadepth)

  # Non-bag points
  nonbagpoints <- data[datadepth < borderdepth, , drop = FALSE]

  # Find mid (center)
  datadict <- split(data.frame(data), datadepth)
  if (length(datadict[[as.character(max(datadepth))]]) == 1) {
    mid <- as.numeric(datadict[[as.character(max(datadepth))]][1, ])
  } else {
    midmean <- colMeans(datadict[[as.character(max(datadepth))]])
    mid <- midmean / sqrt(sum(midmean^2))
  }

  # Generate sphere grid
  phi <- seq(0, pi, length.out = res)
  theta <- seq(0, 2 * pi, length.out = res)
  grid <- expand.grid(phi = phi, theta = theta)
  x_sphere <- sin(grid$phi) * cos(grid$theta)
  y_sphere <- sin(grid$phi) * sin(grid$theta)
  z_sphere <- cos(grid$phi)

  # Rotation matrix
  rotation_matrix <- ortho_matrix(mid)

  # Rotate sphere
  rotated_coords <- t(rotation_matrix %*% t(cbind(x_sphere, y_sphere, z_sphere)))
  x_sphere_rotated <- rotated_coords[, 1]
  y_sphere_rotated <- rotated_coords[, 2]
  z_sphere_rotated <- rotated_coords[, 3]

  # Reshape to matrix
  x_sphere_rotated <- matrix(x_sphere_rotated, nrow = res, ncol = res)
  y_sphere_rotated <- matrix(y_sphere_rotated, nrow = res, ncol = res)
  z_sphere_rotated <- matrix(z_sphere_rotated, nrow = res, ncol = res)

  # Surface colors (depth > borderdepth)
  surface_colors_data <- matrix(0, nrow = res, ncol = res)
  for (i in 1:res) {
    coords <- cbind(x_sphere_rotated[, i], y_sphere_rotated[, i], z_sphere_rotated[, i])
    depths <- ahD(data, weights, coords)
    surface_colors_data[, i] <- as.numeric(depths > borderdepth)
  }

  # Find borders
  borders <- list()
  for (i in 1:res) {
    zero_indices <- which(surface_colors_data[, i] == 0)
    if (length(zero_indices) > 0) {
      border_idx <- zero_indices[1] - 1
      if (border_idx > 0) {
        borders <- c(borders, list(c(border_idx, i)))
      }
    }
  }

  # Calculate distances from mid
  middist <- matrix(0, nrow = res, ncol = res)
  for (i in 1:res) {
    for (j in 1:res) {
      point <- c(x_sphere_rotated[i, j], y_sphere_rotated[i, j], z_sphere_rotated[i, j])
      if (dist == "arc") {
        middist[i, j] <- arcdist(point, mid)
      } else if (dist == "cos") {
        middist[i, j] <- cosdist(point, mid)
      } else if (dist == "chord") {
        middist[i, j] <- chorddist(point, mid)
      } else {
        stop("Wrong format of dist")
      }
    }
  }

  # Border distance
  border_distances <- sapply(borders, function(b) middist[b[1], b[2]])
  if (borderdist == "max") {
    borderd <- max(border_distances)
  } else if (borderdist == "mean") {
    borderd <- mean(border_distances)
  } else {
    stop("Wrong format of borderdist")
  }

  # Factor
  if (dist == "arc") {
    factor <- arcMF(kappaarc(borderd), a)
  } else if (dist == "cos") {
    factor <- cosMF(kappacos(borderd), a)
  } else if (dist == "chord") {
    factor <- chordMF(kappachord(borderd), a)
  } else {
    stop("Wrong format of dist")
  }

  # Count outliers
  num_outliers <- 0
  for (k in 1:nrow(nonbagpoints)) {
    point <- nonbagpoints[k, ]
    grid <- points_on_semicircle(point, mid, res)
    if (nrow(grid) == 0) next
    depths <- ahD(data, weights, grid)
    bagornot <- as.numeric(depths >= borderdepth)
    borderindex <- which(bagornot == 0)[1] - 1
    if (is.na(borderindex) || borderindex < 1) next

    if (dist == "arc") {
      borderd_point <- arcdist(grid[borderindex, ], mid)
      loopd <- factor * borderd_point
      pointdist <- arcdist(point, mid)
    } else if (dist == "cos") {
      borderd_point <- cosdist(grid[borderindex, ], mid)
      loopd <- factor * borderd_point
      pointdist <- cosdist(point, mid)
    } else if (dist == "chord") {
      loopd <- factor * borderd
      pointdist <- chorddist(point, mid)
    } else {
      stop("Wrong format of dist")
    }

    if (pointdist > loopd) {
      num_outliers <- num_outliers + 1
    }
  }

  return(num_outliers)
}
